src/setup/Agda/Setup.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Agda's self-setup.
+
+module Agda.Setup
+  ( getAgdaAppDir
+  , getDataDir
+  , getDataFileName
+  , setup
+  )
+where
+
+import           Control.Monad              ( forM, forM_, when )
+
+import           Data.ByteString            ( ByteString )
+import qualified Data.ByteString            as BS
+import           Data.Functor               ( (<&>) )
+
+import           Language.Haskell.TH.Syntax ( qAddDependentFile, runIO )
+
+-- Import instance Lift ByteString if not supplied by bytestring.
+#if !MIN_VERSION_bytestring(0,11,2)
+import           Instances.TH.Lift          ()
+#endif
+
+import           System.Directory
+  ( XdgDirectory (..)
+  , canonicalizePath, createDirectoryIfMissing, doesDirectoryExist
+  , getAppUserDataDirectory, getXdgDirectory
+  )
+import           System.Environment         ( lookupEnv )
+import           System.FilePath            ( (</>), joinPath, splitFileName )
+
+import           Agda.Setup.DataFiles       ( dataFiles, dataPath )
+import           Agda.VersionCommit         ( versionWithCommitInfo )
+
+-- Tell TH that all the dataFiles are needed for compilation.
+[] <$ mapM_ (qAddDependentFile . dataPath) dataFiles
+
+-- | The embedded contents of the Agda data directory,
+--   generated by Template Haskell at compile time.
+
+embeddedDataDir :: [(FilePath, ByteString)]
+embeddedDataDir = $(do
+
+    -- Load all the dataFiles.
+    contents <- runIO $ mapM (BS.readFile . dataPath) dataFiles
+
+    -- Return the association list as Exp.
+    [| zip dataFiles contents |]
+  )
+
+-- | Get the path to @~/.agda@ (system-specific).
+--   Can be overwritten by the @AGDA_DIR@ environment variable.
+--
+--   (This is not to be confused with the directory 'getDataDir' for the data files
+--   that Agda needs (e.g. the primitive modules).)
+--
+getAgdaAppDir :: IO FilePath
+getAgdaAppDir = do
+  -- The default can be overwritten by setting the AGDA_DIR environment variable
+  lookupEnv "AGDA_DIR" >>= \case
+    Nothing -> agdaDir
+    Just dir -> doesDirectoryExist dir >>= \case
+      True  -> canonicalizePath dir
+      False -> do
+        d <- agdaDir
+        putStrLn $ "Warning: Environment variable AGDA_DIR points to non-existing directory " ++ show dir ++ ", using " ++ show d ++ " instead."
```

Comment:
This is how it used to be (I just copied it), but printing this warning every time one wants to access the AGDA_DIR is arguably broken.

Reply:
Shouldn't we create it, assuming it is non-empty?

Reply:
Not sure.
If `$AGDA_DIR` does not exist, then I would consider it most likely that the user has misspelled it.  Just creating it and continuing will likely produce an unexpected and puzzling behavior.
It seems better to have the user (or script) create this directory explicitly when using the `AGDA_DIR` environment variable.

However, it may be clearest if Agda simply crashes in case `$AGDA_DIR` does not exist, rather than falling back to the system default.  This would prompt the user to either unset the `AGDA_DIR` variable or fix its content.